### PR TITLE
sql: disallow ambiguous table references

### LINKED ISSUE #4756: sql: multiple references to same table in join do not require aliases
In PG....
```
SELECT * FROM a CROSS JOIN a;
```
produces...
```
ERROR:  table name "a" specified more than once
```

However, Materialize allows this without issue.

Postgres checks against for these errors in [`checkNameSpaceConflicts`](https://github.com/postgres/postgres/blob/master/src/backend/parser/parse_relation.c#L414).

I did a cursory timeboxed check, and we needed something more complex than simply checking for duplicate `ScopeItem[0]` names, so filing this away for another moment in time.

### PR DESCRIPTION
**Pull request by sploiselle**
### Motivation

This PR fixes a recognized bug.

Fixes #4756

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR adds a release note for any
  [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/user/content/release-notes.md#what-changes-require-a-release-note).

### CODE REVIEW

**Review by asenac — APPROVED** [review]
Wow, very neat fix.

File: doc/user/content/release-notes.md [hunk 1]
```diff
@@ -103,6 +103,32 @@ List new features before bug fixes.
...
+- **Breaking change**: Disallow ambiguous table references in queries. For
+  example:
+
+  ```sql
+  SELECT * FROM a, a;
+  SELECT * FROM a, b AS a;
+  SELECT * FROM a, generate_series(1,3) AS a;
+  ```
+
+  These queries previously worked, but will both now throw the error:
```
Comment (mjibson) [inline doc/user/content/release-notes.md:115]: "both" is wrong since there's not 2

File: src/sql/src/plan/scope.rs [hunk 2]
```diff
@@ -302,15 +305,24 @@ impl Scope {
             })
     }
 
-    pub fn product(self, right: Self) -> Self {
-        Scope {
+    pub fn product(self, right: Self) -> Result<Self, anyhow::Error> {
+        let l_tables = self.table_names();
+        let r_tables = right.table_names();
+        for l in l_tables {
```
Comment (mjibson) [inline src/sql/src/plan/scope.rs:311]: Is this better than `l_tables.intersection(r_tables)`?

File: test/testdrive/delete-using.td [hunk 3]
```diff
@@ -16,7 +16,8 @@
 > CREATE TABLE t3 (f1 INTEGER);
 
 # Ambiguous table name - currently succeeds due to #4756
```
Comment (mjibson) [inline test/testdrive/delete-using.td:18]: Can delete this comment.

File: test/testdrive/duplicate-table-names.td [hunk 4]
```diff
@@ -0,0 +1,84 @@
+# Copyright Materialize, Inc. and contributors. All rights reserved.
+#
+# Use of this software is governed by the Business Source License
+# included in the LICENSE file at the root of this repository.
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+
+# Copyright 2015 - 2019 The Cockroach Authors. All rights reserved.
```
Comment (mjibson) [inline test/testdrive/duplicate-table-names.td:10]: Is this supposed to be here?

**Review by mjibson — APPROVED** [review]

File: src/sql/src/plan/scope.rs (hunk 2 above)
Comment (mjibson) [reply src/sql/src/plan/scope.rs:311]: Ok. The other annoyance with this for loop is that if there are multiple conflicting names, it's not deterministic which one will print. Not a huge deal, but would be annoying if any tests look for that. Still fine moving forward as is though.

Comment (sploiselle) [reply src/sql/src/plan/scope.rs:311]: Very fair point; if I dip back into refactoring this for Phillip's comments will make the output deterministic.

**Review by philip-stoev — APPROVED** [review]
Thank you very much for the extensive tests! I only added two more queries to the .td files and pushed a new commit on top of your branch -- if you can squash it in  that would be much appreciated.

**Pull request merged**